### PR TITLE
Mergeup zephyr 2.0

### PR DESCRIPTION
**Pull request by mbolivar**
A couple of tweaks to the new/experimental west extensions, mergeups, and patches to make things build again without warnings.

I would appreciate a careful review of the bluetooth changes. I don't really know what I'm doing; this is just based on what upstream did.

### CODE REVIEW

**Comment by mbolivar** [comment]
DNM for now while we're letting CI have its say.

**Comment by mbolivar** [comment]
@thst-nordic is CI stuck? I've been watching for a few hours and no results.

**Comment by pdunaj** [comment]
Hi @mbolivar , we see it too. Unfortunately @thst-nordic is out-of-office as far as I know.

**Comment by pdunaj** [comment]
Hi @wiminordic , do you know what can be wrong?

**Comment by pdunaj** [comment]
```
Checking out Revision 8715d663295a0b8dfcbadcb20646920f5dbfbfca (PR-1135)
Commit message: "samples: bluetooth: add pairing_failed reason parameters"
First time build. Skipping changelog.
Cleaning workspace
java.io.IOException: No space left on device
```

**Comment by pdunaj** [comment]
@mbolivar , @wiminordic , it seems that CI tries to run something but there is no space.

**Comment by mbolivar** [comment]
> @mbolivar , @wiminordic , it seems that CI tries to run something but there is no space.

@pdunaj thanks for the diagnosis. I don't know the servers that are responsible for this, but is there a /tmp or other directory that can be cleared out?

**Comment by mbolivar** [comment]
I'm currently working remotely until next month, and I've had limited luck working with shared drives under the VPN. Not sure if I can help in any way but if so please let me know; I'll be around for a bit longer.

**Comment by pdunaj** [comment]
Hi @mbolivar , I am looking who can fix this but it seems the issue is known at least from 16:00 yesterday and is not fixed.
@carlescufi

**Comment by mbolivar** [comment]
> Hi @mbolivar , I am looking who can fix this but it seems the issue is known at least from 16:00 yesterday and is not fixed.
> @carlescufi

Carles is out this week for a management training. Is there anyone else who understands the CI available?

**Comment by pdunaj** [comment]
Hi @mbolivar , it's being fixed now as far as I know.

**Comment by mbolivar** [comment]
Current failures seem to be due to CI environment updates needed:

```
  The SDK version you are using is too old, please update your SDK.


  You need at least SDK version 0.10.3.

```

**Comment by SebastianBoe** [comment]
I wonder if the need for 0.10.3 is artificial and we could patch zephyr to be OK with 0.10.2.

**Comment by carlescufi** [comment]
> I wonder if the need for 0.10.3 is artificial and we could patch zephyr to be OK with 0.10.2.

I don't think that is worth it. Might as well update the SDK and keep up.

**Comment by carlescufi** [comment]
> Checked with PCA20041. Device connects and seems to be pairing fine but connection fails.

OK, are you using the legacy controller? If so, can you please show @joerchan and @cvinayak the failure and see if we can spot the problem?

**Comment by carlescufi** [comment]
> > Checked with PCA20041. Device connects and seems to be pairing fine but connection fails.
> 
> OK, are you using the legacy controller? If so, can you please show @joerchan and @cvinayak the failure and see if we can spot the problem?

Can you enable the ticker compatibility mode?

**Comment by nvlsianpu** [comment]
From error log:
```
  You need at least SDK version 0.10.3.
  You have version 0.10.2 (/opt/zephyr-sdk).
```

@thst-nordic  need to upgrade sdk version

**Comment by pdunaj** [comment]
@carlescufi , @joerchan , @cvinayak 
We are using legacy controller. I don't see anything specific in the logs. When I enable bt debug logs there is following log and then disconnection

```
[01183137] <dbg> bt_hci_core.bt_hci_cmd_create: opcode 0x0406 param_len 3
[01183138] <dbg> bt_hci_core.bt_hci_cmd_create: buf 0x200094b0
[01183138] <dbg> bt_hci_core.bt_hci_cmd_send: opcode 0x0406 len 6
[01183139] <dbg> bt_conn.bt_conn_set_state: connected -> disconnect
[01183139] <err> ble_state: Security establishment failed - device disconnected
```

**Comment by pdunaj** [comment]
@carlescufi , I have switched to split LL and the effect is the same

File: applications/nrf_desktop/src/modules/ble_state.c [hunk 1]
```diff
@@ -215,15 +222,16 @@ static void security_changed(struct bt_conn *conn, bt_security_t level)
 
 	struct bt_conn_info info;
 
-	int err = bt_conn_get_info(conn, &info);
-	if (err) {
+	int rc = bt_conn_get_info(conn, &info);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_state.c]: @mbolivar Please leave it as is for the sake of consistency. Instead rename the new argument to `bt_err`

**Comment by pdunaj** [comment]
> > > Checked with PCA20041. Device connects and seems to be pairing fine but connection fails.
> > 
> > 
> > OK, are you using the legacy controller? If so, can you please show @joerchan and @cvinayak the failure and see if we can spot the problem?
> 
> Can you enable the ticker compatibility mode?

@carlescufi , I enabled it and this time phone reacted as it would during a connection. Unfortunately the device did not survived the process :(

```
[00863037] <inf> event_manager: e: ble_peer_event id=0x200027c4 SECURED
[00863039] <inf> ble_bond: Erased peer
[00863446] <inf> event_manager: e: ble_peer_operation_event ERASED bt_app_id=0 bt_stack_id=1
[00863448] <inf> event_manager: e: led_event led_id:1 effect:0x3a5e0
[00863451] <inf> event_manager: e: led_event led_id:1 effect:0x3a5c8
[00896446] <inf> ble_state: Conn parameters updated
        interval 0x002
        lat 
        timeout 50

***** BUS FAULT *****  Instruction bus errorr0/a1:  0x00000000  r1/a2:  0x00000001  r2/a3:  0x20001128r3/a4:  0x380003d3 r12/ip:  0x00000001 r14/lr:  0x00024e99 xpsr:  0x61000000Faulting instruction address (r15/pc): 0x380003d2>>> ZEPHYR FATAL ERROR 0: CPU exceptionCurrent thread: 0x20004344 (unknown)FATAL: 
FATAL: 
FATAL: 
FATAL: 
FATAL: 
FATAL: 
FATAL: 
FATAL: 
FATAL: Halting system
```

**Comment by carlescufi** [comment]
@pdunaj I think it might be time for you to switch from the legacy to the split controller. this might require a bit of rework of memory usage on your side, but I believe it's worth it. I've asked @cvinayak to look at this with you.

**Comment by pdunaj** [comment]
Hi @carlescufi , @joerchan ,

I spent some time together with @cvinayak today. The connection problems we see seems to be caused by new ticker and this is a known issue.
As I mentioned earlier when I enable the ticker compatibility device crashes in a nasty way. This is caused by stack overflow. It seems that Zephyr 2.0 requires increase in BLE TX stack. This is however only when logger is disabled. When logger is enabled the RAM consumption is even higher and I had to increase both TX and RX stack (and both by high amount).
So all-in-all we have 2 issues:
- broken ticket
- increased RAM consumption

@carlescufi , we cannot switch to spit link layer until it would be considered mature (at this point in the project it would be a huge risk), I heard it is not.

@jhn-nordic , the increased memory requirements of BLE stack may be a problem for desktop mouse.

**Comment by jhn-nordic** [comment]
FYI: We are on the limit when it comes to our ability to squeeze the nRF52 Desktop into our 24 kB RAM devices. ~98-99% RAM utilization in release mode with no debug options.

**Comment by pdunaj** [comment]
Today I was able to spend a bit of time on this and it seems that the problem with FAULT is not connected with stack sizes. From stack profiling it seems that usage is within limits and even if I set the BLE stacks to 8k the fault occurs (although it occurs not at the first time). 
Maybe some memory is corrupted (e.g. by out-of-bounds write)?

The other thing I noticed is that even if I get the connection there is no HID data transmission. I have not check yet why.

**Comment by pdunaj** [comment]
It seems that problem happens in `bt_gatt_attr_write_ccc` while the bond is being added.

**Comment by kapi-no** [comment]
> 
> 
> It seems that problem happens in `bt_gatt_attr_write_ccc` while the bond is being added.
> 
> handler pointer stored in hid service in ncs is incorrect (0x38000462). My guess is that some macro has changed and code was not updated yet. let's see.

This is right, HIDS module needs some alignments after this mergup. I think there is no ```ccc``` field anymore in the ```struct _bt_gatt_ccc``` structure. The CCC changed callback logic can fail because of this. Example:

https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/master/subsys/bluetooth/services/hids.c#L464

File: include/bluetooth/gatt_pool.h [hunk 2]
```diff
@@ -104,11 +104,10 @@ extern "C" {
 #define BT_GATT_POOL_CCC(_gp, _ccc_cfg, _ccc_cb)                               \
 	do {                                                                   \
 		int _ret;                                                      \
-		const struct _bt_gatt_ccc _ccc = {                             \
-		    .cfg = _ccc_cfg,                                           \
-		    .cfg_len = ARRAY_SIZE(_ccc_cfg),                           \
+		struct _bt_gatt_ccc _ccc = {                                   \
 		    .cfg_changed = _ccc_cb,                                    \
 		};                                                             \
+		memcpy(&_ccc.cfg, (_ccc_cfg), sizeof(_ccc.cfg));               \
```
Comment (pdunaj) [inline include/bluetooth/gatt_pool.h]: this is incorrect - the hids relies on cfg to be in the original structure.

**Comment by pdunaj** [comment]
Unfortunately until the problem with `BT_GATT_POOL_CCC` is fixed this PR cannot be merged. 
@mbolivar , @jhn-nordic , @carlescufi

**Comment by nvlsianpu** [comment]
Can we depopulate mcuboot update from there?
Her is the manifest patch https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1155

**Comment by joerchan** [comment]
@pdunaj @kapi-no 
Fix for the HIDS report context for CCCD  changed handler is here: https://github.com/mbolivar/fw-nrfconnect-nrf/commit/8312c52b54114d3a893bd2692bb095f8b78805d0
This one is dependant on change to the GATT api, cherry-pick into the upmerge for zephyr.

**Comment by pdunaj** [comment]
@mbolivar , could you resolve conflicts and force push again? Does CI have the new SDK version in place?

**Comment by mbolivar** [comment]
> @mbolivar , could you resolve conflicts and force push again? Does CI have the new SDK version in place?

The SDK was updated and I've rebased, so we can keep going from here.

**Comment by mbolivar** [comment]
k

> Can we depopulate mcuboot update from there?
> Her is the manifest patch #1155

Great, thanks for that. I've removed the mcuboot portion of the west.yml changes.

**Comment by pdunaj** [comment]
@wiminordic , it seems that the test failed again. Is this real problem or something with the test?

**Comment by pdunaj** [comment]
> @wiminordic , it seems that the test failed again. Is this real problem or something with the test?

Obviously it will not work as ticker compatibility is not enabled... I will enabled it and push the commit to this PR.

**Comment by wiminordic** [comment]
@pdunaj @mbolivar The issue is that Service Discovery takes now significantly more time (30s) at the same time CPU is timing out and we get a disconnect. It will be fixed in the test but why is service discovery taking so much time now?

**Comment by pdunaj** [comment]
> @pdunaj @mbolivar The issue is that Service Discovery takes now significantly more time (30s) at the same time CPU is timing out and we get a disconnect. It will be fixed in the test but why is service discovery taking so much time now?

Thanks for checking this. The problem can come from ticker. There is new implementation of it in Zephyr 2.0 that does not work. It is however default one so by default we have problems with connection. If ticker compatibility is enabled link seems to work fine. I have pushed the commit enabling ticker compatibility **in our project** only. I expect our tests will pass. I tested the device with Android and it worked. I will check now dongle but I expect it to work too.

I expect that BLE samples will stop working as they don't have ticker compatibility enabled. If we have anything running connection tests in them in CI we should see it. (pinging @grochu just in case)

**Comment by pdunaj** [comment]
Hi @wiminordic , can you confirm that CI run and pass dongle tests? I tried device/dongle connection just now and I see problems. The BLE seems to work fine but data does not pass. The problem may be with the USB this time. I see following in the logs:
```
[00018221] <wrn> usb_device: Failed to enable endpoint 81
```
We got callbacks, set USB state to active and enable notifications but data is not transmitted.

**Comment by pdunaj** [comment]
Yes, USB transmission does not occur.

File: include/bluetooth/gatt_pool.h [hunk 3]
```diff
@@ -98,16 +98,14 @@ extern "C" {
...
  */
-#define BT_GATT_POOL_CCC(_gp, _ccc_cfg, _ccc_cb)                               \
+#define BT_GATT_POOL_CCC(_gp, _ccc, _ccc_changed)                              \
 	do {                                                                   \
 		int _ret;                                                      \
-		struct _bt_gatt_ccc _ccc = {                                   \
-		    .cfg_changed = _ccc_cb,                                    \
-		};                                                             \
-		memcpy(&_ccc.cfg, (_ccc_cfg), sizeof(_ccc.cfg));               \
+		_ccc = (struct _bt_gatt_ccc)BT_GATT_CCC_INITIALIZER(\
```
Comment (rakons) [inline include/bluetooth/gatt_pool.h]: (_ccc) = ....

File: include/bluetooth/gatt_pool.h [hunk 4]
```diff
@@ -98,16 +98,14 @@ extern "C" {
...
+			_ccc_changed, NULL, NULL);      \
 		_ret = bt_gatt_pool_ccc_alloc(_gp, &_ccc);                     \
```
Comment (rakons) [inline include/bluetooth/gatt_pool.h]: (_gp), &(_ccc)

File: include/bluetooth/gatt_pool.h [hunk 5]
```diff
@@ -98,16 +98,14 @@ extern "C" {
...
-#define BT_GATT_POOL_CCC(_gp, _ccc_cfg, _ccc_cb)                               \
+#define BT_GATT_POOL_CCC(_gp, _ccc, _ccc_changed)                              \
 	do {                                                                   \
 		int _ret;                                                      \
-		struct _bt_gatt_ccc _ccc = {                                   \
-		    .cfg_changed = _ccc_cb,                                    \
-		};                                                             \
-		memcpy(&_ccc.cfg, (_ccc_cfg), sizeof(_ccc.cfg));               \
+		_ccc = (struct _bt_gatt_ccc)BT_GATT_CCC_INITIALIZER(\
+			_ccc_changed, NULL, NULL);      \
```
Comment (rakons) [inline include/bluetooth/gatt_pool.h]: (_ccc_changed)

File: subsys/bluetooth/gatt_pool.c [hunk 6]
```diff
@@ -414,7 +379,7 @@ int bt_gatt_pool_desc_alloc(struct bt_gatt_pool *gp,
 }
 
 int bt_gatt_pool_ccc_alloc(struct bt_gatt_pool *gp,
-			   struct _bt_gatt_ccc const *ccc)
+			   struct _bt_gatt_ccc *ccc)
```
Comment (rakons) [inline subsys/bluetooth/gatt_pool.c]: Why this cannot be const anymore? I cannot see a reason directly in this code changes.

**Comment by pdunaj** [comment]
I moved back to the old revision (`0aa796cbaa917d1f2bcb3fe2b5b8d9abda206a26`) on the USB subsystem and the communication works fine. It seems there is some problem with how our driver handles state changes and what stack sets. I see there are at least several changes that could affect the behavior.
The problem replicates only on the dongle - mouse works find on both BLE and USB.

**Comment by pdunaj** [comment]
I am leaving the investigation here as I cannot focus on fixing the USB.

@wiminordic before this PR is merged we need to run our full tests to see that USB issue is indeed fixed.

**Comment by kapi-no** [comment]
> I expect that BLE samples will stop working as they don't have ticker compatibility enabled. If we have anything running connection tests in them in CI we should see it. (pinging @grochu just in case)

I tested HIDS Peripheral mouse sample and it works without any KConfig modifications (the sample was able to maintain the connection for 30 mins+).

File: subsys/bluetooth/gatt_pool.c (hunk 6 above)
Comment (joerchan) [reply subsys/bluetooth/gatt_pool.c]: Previously we made a copy of this one, and assigned attr->user_data to point to the copy, now it points to this object. It has to be modifiable now.

**Comment by mbolivar** [comment]
@pdunaj just FYI, @carlescufi has requested that we merge this and fix the remaining issues afterwards, so I will be doing that unless any additional unexpected surprises result in this latest CI run.

**Comment by pdunaj** [comment]
> @pdunaj just FYI, @carlescufi has requested that we merge this and fix the remaining issues afterwards, so I will be doing that unless any additional unexpected surprises result in this latest CI run.

@mbolivar This is not acceptable. The nrf repository is, for good and bad, a home for multiple project. You cannot simply merge a PR if you see it was blocked. Especially if explanation was given. The simple reason for it is that you may not be aware of implications of your action on other projects schedule.

I have sent an e-mail to @carlescufi to clarify this. I am especially surprised it happened considering how careful merging rights are given. It could be a misunderstanding, but if so instructions in Confluence are not clear enough.

The other question is about the bugs introduced by this merge. We need to confirm if these changes can stay in and we can wait for the USB fix or this PR should be reverted. (@jhn-nordic )

**Comment by carlescufi** [comment]
> > @pdunaj just FYI, @carlescufi has requested that we merge this and fix the remaining issues afterwards, so I will be doing that unless any additional unexpected surprises result in this latest CI run.
> 
> @mbolivar This is not acceptable. The nrf repository is, for good and bad, a home for multiple project. You cannot simply merge a PR if you see it was blocked. Especially if explanation was given. The simple reason for it is that you may not be aware of implications of your action on other projects schedule.
> 
> I have sent an e-mail to @carlescufi to clarify this. I am especially surprised it happened considering how careful merging rights are given. It could be a misunderstanding, but if so instructions in Confluence are not clear enough.
> 
> The other question is about the bugs introduced by this merge. We need to confirm if these changes can stay in and we can wait for the USB fix or this PR should be reverted. (@jhn-nordic )

Thanks for the insight Pawel.
I have now revisited this in an internal email thread, let's keep the discussion there.

**Pull request merged**